File: worker/src/rag_worker/repository.py
from datetime import datetime, timedelta, timezone
from logging import Logger
from typing import TYPE_CHECKING, Any

from rag_worker.config import WorkerSettings
from rag_worker.types import DocumentRecord, IngestionJob, PreparedChunkRecord

if TYPE_CHECKING:
    from supabase import Client
else:  # pragma: no cover - static typing fallback
    Client = Any


def _utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _stale_lock_cutoff_iso(timeout_seconds: int) -> str:
    return (datetime.now(timezone.utc) - timedelta(seconds=timeout_seconds)).isoformat()


def _safe_str(value: object) -> str:
    return str(value) if value is not None else ""


class WorkerRepository:
    def __init__(self, settings: WorkerSettings, logger: Logger) -> None:
        self.settings = settings
        self.logger = logger
        self.client: Client = self._create_client(settings.supabase_url, settings.supabase_service_role_key)

    @staticmethod
    def _create_client(url: str, service_role_key: str) -> Client:
        try:
            from supabase import create_client
        except ModuleNotFoundError as exc:  # pragma: no cover - runtime environment error
            raise RuntimeError("supabase package is required to run the ingestion worker") from exc

        return create_client(url, service_role_key)

    def claim_next_job(self) -> IngestionJob | None:
        response = (
            self.client.table("ingestion_jobs")
            .select("id,document_id,status,attempt")
            .in_("status", ["queued", "failed"])
            .lt("attempt", self.settings.max_retries)
            .is_("locked_at", "null")
            .order("created_at", desc=False)
            .limit(10)
            .execute()
        )

        for row in response.data or []:
            claimed = (
                self.client.table("ingestion_jobs")
                .update(
                    {
                        "status": "processing",
                        "attempt": int(row["attempt"]) + 1,
                        "locked_at": _utc_now_iso(),
                        "locked_by": self.settings.worker_name,
                    }
                )
                .eq("id", row["id"])
                .eq("status", row["status"])
                .is_("locked_at", "null")
                .execute()
            )

            if claimed.data:
                selected = claimed.data[0]
                return IngestionJob(
                    id=_safe_str(selected["id"]),
                    document_id=_safe_str(selected["document_id"]),
                    status=_safe_str(selected["status"]),
                    attempt=int(selected["attempt"]),
                )

        if self.settings.worker_lock_timeout_seconds <= 0:
            return None

        stale_cutoff = _stale_lock_cutoff_iso(self.settings.worker_lock_timeout_seconds)
        stale_response = (
            self.client.table("ingestion_jobs")
            .select("id,document_id,status,attempt")
            .eq("status", "processing")
            .lt("attempt", self.settings.max_retries)
            .lte("locked_at", stale_cutoff)
            .order("locked_at", desc=False)
            .limit(10)
            .execute()
        )

        for row in stale_response.data or []:
            reclaimed = (
                self.client.table("ingestion_jobs")
                .update(
                    {
                        "status": "processing",
                        "attempt": int(row["attempt"]) + 1,
                        "locked_at": _utc_now_iso(),
                        "locked_by": self.settings.worker_name,
                    }
                )
                .eq("id", row["id"])
                .eq("status", "processing")
                .lte("locked_at", stale_cutoff)
                .execute()
            )

            if reclaimed.data:
                selected = reclaimed.data[0]
                self.logger.warning(
                    "reclaimed_stale_job_lock",
                    extra={
                        "job_id": _safe_str(selected["id"]),
                        "document_id": _safe_str(selected["document_id"]),
                    },
                )
                return IngestionJob(
                    id=_safe_str(selected["id"]),
                    document_id=_safe_str(selected["document_id"]),
                    status=_safe_str(selected["status"]),
                    attempt=int(selected["attempt"]),
                )

        return None

    def get_document(self, document_id: str) -> DocumentRecord:
        response = (
            self.client.table("documents")
            .select("id,storage_path,sha256,title,language,status,ingestion_version")
            .eq("id", document_id)
            .single()
            .execute()
        )

        data = response.data
        if not data:
            raise RuntimeError(f"Document not found: {document_id}")

        language_value = data.get("language")
        if isinstance(language_value, str) and language_value in {"EN", "DE", "FR", "IT", "ES"}:
            language = language_value
        else:
            language = None

        return DocumentRecord(
            id=_safe_str(data["id"]),
            storage_path=_safe_str(data["storage_path"]),
            sha256=_safe_str(data["sha256"]),
            title=data.get("title"),
            language=language,
            status=_safe_str(data["status"]),
            ingestion_version=int(data["ingestion_version"]),
        )

    def download_document(self, storage_path: str) -> bytes:
        payload = self.client.storage.from_(self.settings.rag_storage_bucket).download(storage_path)
        if isinstance(payload, bytes):
            return payload
        if hasattr(payload, "content"):
            return payload.content  # type: ignore[return-value]
        raise RuntimeError("Unexpected storage download payload")

    def set_document_status(self, document_id: str, status: str, language: str | None = None) -> None:
        update_payload: dict[str, object] = {"status": status}
        if language:
            update_payload["language"] = language

        self.client.table("documents").update(update_payload).eq("id", document_id).execute()

    def replace_document_chunks(self, document_id: str, chunks: list[PreparedChunkRecord]) -> None:
        self.client.table("document_chunks").delete().eq("document_id", document_id).execute()

        if not chunks:
            return

        batch_size = max(1, self.settings.worker_chunk_insert_batch_size)

        for index in range(0, len(chunks), batch_size):
            batch = chunks[index : index + batch_size]
            rows = [
                {
                    "document_id": item.document_id,
                    "chunk_index": item.chunk_index,
                    "page_number": item.page_number,
                    "section_title": item.section_title,
                    "content": item.content,
                    "context": item.context,
                    "language": item.language,
                    "embedding": item.embedding,
                }
                for item in batch
            ]
            self.client.table("document_chunks").insert(rows).execute()

    def mark_job_completed(self, job_id: str) -> None:
        (
            self.client.table("ingestion_jobs")
            .update(
                {
                    "status": "completed",
                    "last_error": None,
                    "locked_at": None,
                    "locked_by": None,
                }
            )
            .eq("id", job_id)
            .execute()
        )

    def mark_job_failed(self, job: IngestionJob, error_message: str) -> bool:
        dead_letter = job.attempt >= self.settings.max_retries
        status = "dead_letter" if dead_letter else "failed"

        (
            self.client.table("ingestion_jobs")
            .update(
                {
                    "status": status,
                    "last_error": error_message[:4000],
                    "locked_at": None,
                    "locked_by": None,
                }
            )
            .eq("id", job.id)
            .execute()
        )

        return dead_letter

File: lib/ingestion/runtime/repository.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import type {
  ChunkCandidate,
  DocumentRecord,
  IngestionJob,
  IngestionRuntimeSettings,
  JobProgress,
  PreparedChunkRecord,
  RuntimeLogger,
} from "@/lib/ingestion/runtime/types";
import { getSupabaseAdminClient } from "@/lib/supabase/admin";
import type { Database, SupportedLanguage } from "@/lib/supabase/database.types";

type ClaimedJobRow = Database["public"]["Functions"]["claim_ingestion_jobs"]["Returns"][number];
type DocumentRow = Database["public"]["Tables"]["documents"]["Row"];
type ReplaceDocumentChunksRow = Database["public"]["Functions"]["replace_document_chunks"]["Returns"][number];
type AppendDocumentChunksRow = Database["public"]["Functions"]["append_document_chunks"]["Returns"][number];

export type ClaimIngestionJobsInput = {
  workerName: string;
  batchSize: number;
  lockTimeoutSeconds: number;
  maxRetries?: number;
};

export interface IngestionRuntimeRepository {
  claimIngestionJobs(input: ClaimIngestionJobsInput): Promise<IngestionJob[]>;
  getDocument(documentId: string): Promise<DocumentRecord>;
  downloadDocument(storagePath: string): Promise<Uint8Array>;
  replaceDocumentChunks(documentId: string, chunks: PreparedChunkRecord[]): Promise<void>;
  markJobCompleted(jobId: string, language?: SupportedLanguage | null): Promise<void>;
  markJobFailed(job: IngestionJob, errorMessage: string): Promise<boolean>;
  invalidateRetrievalCache(): Promise<void>;
  saveChunkCandidates(jobId: string, chunks: ChunkCandidate[], total: number): Promise<void>;
  loadJobProgress(jobId: string): Promise<JobProgress>;
  updateJobStage(jobId: string, stage: string): Promise<void>;
  updateJobProgress(jobId: string, chunksProcessed: number): Promise<void>;
  yieldJob(jobId: string): Promise<void>;
  insertChunkBatch(documentId: string, chunks: PreparedChunkRecord[]): Promise<void>;
}

function isMissingRpcFunction(errorMessage: string): boolean {
  return errorMessage.includes("Could not find the function");
}

function buildMissingRpcError(functionName: string, details?: string): Error {
  const suffix = details ? ` (${details})` : "";
  return new Error(`Required ingestion RPC ${functionName} is unavailable${suffix}`);
}

function toIngestionJob(row: ClaimedJobRow): IngestionJob {
  return {
    id: row.id,
    documentId: row.document_id,
    status: row.status,
    attempt: row.attempt,
    currentStage: "claimed",
  };
}

function toDocumentRecord(row: DocumentRow): DocumentRecord {
  return {
    id: row.id,
    storagePath: row.storage_path,
    sha256: row.sha256,
    title: row.title,
    language: row.language,
    status: row.status,
    ingestionVersion: row.ingestion_version,
  };
}

function toReplaceDocumentChunkRpcChunk(chunk: PreparedChunkRecord): Record<string, unknown> {
  return {
    chunk_index: chunk.chunkIndex,
    page_number: chunk.pageNumber,
    section_title: chunk.sectionTitle,
    content: chunk.content,
    context: chunk.context,
    language: chunk.language,
    embedding: chunk.embedding,
  };
}

async function toUint8Array(payload: unknown): Promise<Uint8Array> {
  if (!payload) {
    throw new Error("Storage download returned empty payload");
  }

  if (payload instanceof Uint8Array) {
    return payload;
  }

  if (payload instanceof ArrayBuffer) {
    return new Uint8Array(payload);
  }

  if (typeof Blob !== "undefined" && payload instanceof Blob) {
    return new Uint8Array(await payload.arrayBuffer());
  }

  if (Buffer.isBuffer(payload)) {
    return new Uint8Array(payload);
  }

  if (typeof payload === "object" && "arrayBuffer" in payload) {
    const candidate = payload as { arrayBuffer: () => Promise<ArrayBuffer> };
    return new Uint8Array(await candidate.arrayBuffer());
  }

  throw new Error("Unsupported storage download payload");
}

export class SupabaseIngestionRuntimeRepository implements IngestionRuntimeRepository {
  private readonly supabase: SupabaseClient<Database>;
  private readonly settings: IngestionRuntimeSettings;
  private readonly logger: RuntimeLogger;

  constructor(input: {
    settings: IngestionRuntimeSettings;
    logger: RuntimeLogger;
    supabase?: SupabaseClient<Database>;
  }) {
    this.settings = input.settings;
    this.logger = input.logger;
    this.supabase = input.supabase ?? getSupabaseAdminClient();
  }

  async claimIngestionJobs(input: ClaimIngestionJobsInput): Promise<IngestionJob[]> {
    const { data, error: rpcError } = await this.supabase.rpc("claim_ingestion_jobs", {
      worker_name: input.workerName,
      batch_size: Math.max(1, input.batchSize),
      lock_timeout_seconds: Math.max(1, input.lockTimeoutSeconds),
      max_retries: input.maxRetries ?? this.settings.maxRetries,
    });

    if (!rpcError) {
      return (data ?? []).map(toIngestionJob);
    }

    if (isMissingRpcFunction(rpcError.message)) {
      throw buildMissingRpcError("claim_ingestion_jobs", rpcError.message);
    }

    throw new Error(`Failed to claim ingestion jobs via RPC: ${rpcError.message}`);
  }

  async getDocument(documentId: string): Promise<DocumentRecord> {
    const { data, error } = await this.supabase
      .from("documents")
      .select("id,storage_path,sha256,title,language,status,ingestion_version")
      .eq("id", documentId)
      .single<DocumentRow>();

    if (error) {
      throw new Error(`Failed to load document: ${error.message}`);
    }

    return toDocumentRecord(data);
  }

  async downloadDocument(storagePath: string): Promise<Uint8Array> {
    const { data, error } = await this.supabase.storage
      .from(this.settings.ragStorageBucket)
      .download(storagePath);

    if (error) {
      throw new Error(`Failed to download document from storage: ${error.message}`);
    }

    return toUint8Array(data);
  }

  async replaceDocumentChunks(documentId: string, chunks: PreparedChunkRecord[]): Promise<void> {
    const { data, error: rpcError } = await this.supabase.rpc("replace_document_chunks", {
      target_document_id: documentId,
      target_chunks: chunks.map(toReplaceDocumentChunkRpcChunk),
    });

    if (!rpcError) {
      const row = data?.[0] as ReplaceDocumentChunksRow | undefined;
      if (!row) {
        throw new Error(`replace_document_chunks returned no row for ${documentId}`);
      }
      return;
    }

    if (isMissingRpcFunction(rpcError.message)) {
      throw buildMissingRpcError("replace_document_chunks", rpcError.message);
    }

    throw new Error(`Failed to replace document chunks via RPC: ${rpcError.message}`);
  }

  async markJobCompleted(jobId: string, language?: SupportedLanguage | null): Promise<void> {
    const { error: rpcError } = await this.supabase.rpc("complete_ingestion_job", {
      job_id: jobId,
      document_language: language ?? null,
    });

    if (!rpcError) {
      return;
    }

    if (isMissingRpcFunction(rpcError.message)) {
      throw buildMissingRpcError("complete_ingestion_job", rpcError.message);
    }

    throw new Error(`Failed to complete ingestion job via RPC: ${rpcError.message}`);
  }

  async markJobFailed(job: IngestionJob, errorMessage: string): Promise<boolean> {
    const { data, error: rpcError } = await this.supabase.rpc("fail_ingestion_job", {
      job_id: job.id,
      error_text: errorMessage,
      max_retries: this.settings.maxRetries,
    });

    if (!rpcError) {
      const row = data?.[0];
      if (!row) {
        this.logger.warn("ingestion_job_fail_noop", { jobId: job.id });
        return false;
      }
      return row.dead_letter;
    }

    if (isMissingRpcFunction(rpcError.message)) {
      throw buildMissingRpcError("fail_ingestion_job", rpcError.message);
    }

    throw new Error(`Failed to fail ingestion job via RPC: ${rpcError.message}`);
  }

  async invalidateRetrievalCache(): Promise<void> {
    const { data, error: rpcError } = await this.supabase.rpc("invalidate_retrieval_cache");

    if (!rpcError) {
      if (!data?.[0]) {
        throw new Error("invalidate_retrieval_cache returned no row");
      }
      return;
    }

    if (isMissingRpcFunction(rpcError.message)) {
      throw buildMissingRpcError("invalidate_retrieval_cache", rpcError.message);
    }

    throw new Error(`Failed to invalidate retrieval cache via RPC: ${rpcError.message}`);
  }

  async saveChunkCandidates(jobId: string, chunks: ChunkCandidate[], total: number): Promise<void> {
    const { error: rpcError } = await this.supabase.rpc("checkpoint_ingestion_job", {
      target_job_id: jobId,
      target_chunk_candidates: chunks as unknown as Record<string, unknown>[],
      target_chunks_total: total,
      target_chunks_processed: 0,
      target_stage: "chunked",
    });

    if (!rpcError) {
      return;
    }

    if (isMissingRpcFunction(rpcError.message)) {
      throw buildMissingRpcError("checkpoint_ingestion_job", rpcError.message);
    }

    throw new Error(`Failed to save chunk candidates via RPC: ${rpcError.message}`);
  }

  async loadJobProgress(jobId: string): Promise<JobProgress> {
    const { data, error } = await this.supabase
      .from("ingestion_jobs")
      .select("chunk_candidates,chunks_total,chunks_processed,current_stage")
      .eq("id", jobId)
      .single();

    if (error) {
      throw new Error(`Failed to load job progress: ${error.message}`);
    }

    const candidates = data.chunk_candidates
      ? (data.chunk_candidates as unknown as ChunkCandidate[])
      : null;

    return {
      candidates,
      chunksProcessed: data.chunks_processed,
      chunksTotal: data.chunks_total,
      currentStage: data.current_stage,
    };
  }

  async updateJobStage(jobId: string, stage: string): Promise<void> {
    const { error: rpcError } = await this.supabase.rpc("checkpoint_ingestion_job", {
      target_job_id: jobId,
      target_stage: stage,
    });

    if (!rpcError) {
      return;
    }

    if (isMissingRpcFunction(rpcError.message)) {
      throw buildMissingRpcError("checkpoint_ingestion_job", rpcError.message);
    }

    throw new Error(`Failed to update job stage via RPC: ${rpcError.message}`);
  }

  async updateJobProgress(jobId: string, chunksProcessed: number): Promise<void> {
    const { error: rpcError } = await this.supabase.rpc("checkpoint_ingestion_job", {
      target_job_id: jobId,
      target_chunks_processed: chunksProcessed,
    });

    if (!rpcError) {
      return;
    }

    if (isMissingRpcFunction(rpcError.message)) {
      throw buildMissingRpcError("checkpoint_ingestion_job", rpcError.message);
    }

    throw new Error(`Failed to update job progress via RPC: ${rpcError.message}`);
  }

  async yieldJob(jobId: string): Promise<void> {
    const { error: rpcError } = await this.supabase.rpc("yield_ingestion_job", {
      target_job_id: jobId,
    });

    if (!rpcError) {
      return;
    }

    if (isMissingRpcFunction(rpcError.message)) {
      throw buildMissingRpcError("yield_ingestion_job", rpcError.message);
    }

    throw new Error(`Failed to yield job via RPC: ${rpcError.message}`);
  }

  async insertChunkBatch(documentId: string, chunks: PreparedChunkRecord[]): Promise<void> {
    if (chunks.length === 0) {
      return;
    }

    const { data, error: rpcError } = await this.supabase.rpc("append_document_chunks", {
      target_document_id: documentId,
      target_chunks: chunks.map(toReplaceDocumentChunkRpcChunk),
    });

    if (!rpcError) {
      const row = data?.[0] as AppendDocumentChunksRow | undefined;
      if (!row) {
        throw new Error(`append_document_chunks returned no row for ${documentId}`);
      }
      return;
    }

    if (isMissingRpcFunction(rpcError.message)) {
      throw buildMissingRpcError("append_document_chunks", rpcError.message);
    }

    throw new Error(`Failed to append document chunks via RPC: ${rpcError.message}`);
  }
}
